src/workerd/api/sockets.c++

```diff
@@ -128,4 +134,38 @@ void Socket::handleProxyStatus(
   result.markAsHandled();
 }
 
+void Socket::handleReadableEof(jsg::Lock& js) {
+  // Listen for EOF on the ReadableStream.
+  KJ_ASSERT_NONNULL(readable->eofResolverPair).promise.then(js,
+      JSG_VISITABLE_LAMBDA((ref=JSG_THIS), (ref), (jsg::Lock& js) {
+    return ref->maybeCloseWriteSide(js);
```

Comment:
It appears that this doesn't do anything if allowHalfOpen is true. Could we skip waiting altogether in that case?